Mathlib/Lean/Meta/DiscrTree.lean

```diff
@@ -53,19 +53,25 @@ partial def getSubexpressionMatches (d : DiscrTree α s) (e : Expr) : MetaM (Arr
 
 variable {m : Type → Type} [Monad m]
 
-/-- Apply a monadic function to the array of values at each node in a `DiscrTree`. -/
-partial def Trie.mapArraysM (t : DiscrTree.Trie α s) (f : Array α → m (Array β)) :
-    m (DiscrTree.Trie β s) := do
-  match t with
-  | .node vs children =>
-    return .node (← f vs) (← children.mapM fun (k, t') => do pure (k, ← t'.mapArraysM f))
 
-/-- Apply a monadic function to the array of values at each node in a `DiscrTree`. -/
-def mapArraysM (d : DiscrTree α s) (f : Array α → m (Array β)) : m (DiscrTree β s) := do
-  pure { root := ← d.root.mapM (fun t => t.mapArraysM f) }
+/-- A representation of lists-of-quadruples with a smaller memory footprint. -/
+private inductive Ctxt (α β γ δ : Type)
+  | empty : Ctxt α β γ δ
+  | ctxt : α → β → γ → δ → Ctxt α β γ δ → Ctxt α β γ δ
+
+/-- Apply a function to the array of values at each node in a `DiscrTree`. -/
+partial def Trie.mapArrays (t : Trie α s) (f : Array α → Array β) : Trie β s :=
+  let .node vs0 cs0 := t
+  go Ctxt.empty #[] (f vs0) cs0.toList
+where
+  /-- This implementation as a single tail-recursive function is chosen to not blow the
+      interpreter stack when the `Trie` is very deep -/
+  go  | .empty, cs, vs, []                   => .node vs cs
+      | .ctxt cs vs todo k ps, cs', vs', []  => go ps (cs.push (k, .node vs' cs')) vs todo
+      | ps, cs, vs, (k, .node vs' cs')::todo => go (.ctxt cs vs todo k ps) #[] (f vs') cs'.toList
```

Comment:
```suggestion
/-- A stack representation for `Trie.mapArrays` with a smaller memory footprint. -/
private inductive Ctxt {α β s}
  | empty : Ctxt
  | ctxt : Array (Key s × Trie β s) → Array β → Array (Key s × Trie α s) → Nat → Key s → Ctxt → Ctxt

/-- Apply a function to the array of values at each node in a `DiscrTree`. -/
partial def Trie.mapArrays (t : Trie α s) (f : Array α → Array β) : Trie β s :=
  let .node vs0 cs0 := t
  go Ctxt.empty #[] (f vs0) cs0 0
where
  /-- This implementation as a single tail-recursive function is chosen to not blow the
      interpreter stack when the `Trie` is very deep -/
  go ps cs vs todo i := 
    if h : i < todo.size then
      let (k, .node vs' cs') := todo[i]
      go (.ctxt cs vs todo (i+1) k ps) #[] (f vs') cs' 0
    else
      match ps with
      | .empty => .node vs cs
      | .ctxt cs' vs' todo i k ps => go ps (cs'.push (k, .node vs cs)) vs' todo i
```
```

Reply:
You specialize the list-of-quadruples. I considered it, but to me it made the code less readable (it's so verbose), but maybe explicit is better. I assume performance wise it doesn't matter, does it?

I am wary about keeping the full array alive, this means the map operation will need twice the memory, because the old trie can't be freed until the very end. (One could empty out the fields in the old array… but I'm leaning towards elegance here, hoping that the cs array is typically very small anyways - most of the time it's a singleton anyways.)

Reply:
> I assume performance wise it doesn't matter, does it?

It can matter, if any of the fields are scalars. Here I don't think it makes a difference (although `i` could be a scalar if we used `USize` instead of `Nat`).

> You specialize the list-of-quadruples. I considered it, but to me it made the code less readable (it's so verbose), but maybe explicit is better.

I think it makes it clear that this is really specific to this algorithm and is not intended for general use. Also the complete lack of types in this rather complex recursive function is not conducive to understanding. I think `go` should probably just have a full type.

> I am wary about keeping the full array alive, this means the map operation will need twice the memory, because the old trie can't be freed until the very end.

Converting to a list uses more than double the memory already. The main memory savings here are not from the array itself but from the linked sub-tree which could be arbitrarily large, which is why clearing the array elements after we are finished with them is a good idea. But for the currently iterating array, we don't have any choice other than to have a new array and the old array and copy things over. (Actually there is an even more unsafe trick one could play, which is to directly write the `β` elements into the `Array α`, avoiding a reallocation. But this requires dealing with half-typed arrays which is messy.)

Reply:
>  Converting to a list uses more than double the memory already. 

Only for the spine, and only in the nodes under processing, which is much less than the overall data structure. Which I think what we are both saying :-)
It’s important to release the subtree elements (which the naive array-indexing-code didn’t do yet, but either the unsafe overriding or the reversing would do).